File: src/agentsfleetd/http/handlers/runner/heartbeat.zig
//! POST /v1/runners/me/heartbeats — runner liveness + policy delivery.
//!
//! Authed by `runnerBearer` (the principal carries `runner_id`). Every reply
//! carries the row's current assigned policy and degraded verdict, so an
//! operator's dashboard change reaches the host within one heartbeat and a
//! host visit is never needed. `drain`/`stop` statuses arrive with the
//! fleet-failover slice. Side effect: bump `fleet.runners.last_seen_at`
//! (liveness is written here, not on every authed call, per docs/AUTH.md).

const std = @import("std");
const constants = @import("common");
const sql = @import("sql.zig");
const clock = constants.clock;
const logging = @import("log");
const httpz = @import("httpz");
const pg = @import("pg");

const hx_mod = @import("../hx.zig");
const common = @import("../common.zig");
const ec = @import("../../../errors/error_registry.zig");
const protocol = @import("contract").protocol;
const metrics_runner = @import("../../../observability/metrics_runner.zig");
const id_format = @import("../../../types/id_format.zig");
const runner_events = @import("../../../fleet/runner_events.zig");
const policy_row = @import("assigned_policy_row.zig");
const reconcile = @import("heartbeat_reconcile.zig");
const PgQuery = @import("../../../db/pg_query.zig").PgQuery;

const Hx = hx_mod.Hx;
const log = logging.scoped(.runner_heartbeat);
const LOG_EVENT_HEARTBEAT_BUMP_FAILED = "heartbeat_bump_failed";
const LOG_EVENT_CAPABILITY_PERSIST_FAILED = "capability_persist_failed";
const LOG_EVENT_SELFTEST_PERSIST_FAILED = "selftest_persist_failed";
const LOG_EVENT_SELFTEST_REJECTED = "selftest_report_rejected";

pub fn innerRunnerHeartbeat(hx: Hx, req: *httpz.Request) void {
    const runner_id = hx.principal.runner_id orelse {
        // runnerBearer guarantees this is set; defensive only.
        hx.fail(ec.ERR_RUN_INVALID_RUNNER_TOKEN, "runner identity required");
        return;
    };
    const incoming = switch (parseCapabilityReport(hx, req)) {
        .responded => return,
        .none => @as(?protocol.CapabilityReport, null),
        .report => |r| @as(?protocol.CapabilityReport, r),
    };
    const conn = hx.ctx.pool.acquire() catch {
        common.internalDbUnavailable(hx.res, hx.req_id);
        return;
    };
    defer hx.ctx.pool.release(conn);

    // The policy read is load-bearing (the reply is the delivery channel), so
    // it fails loud — the runner retries and keeps its last-applied policy for
    // in-flight work. The verdict + liveness writes below stay best-effort.
    const row = readPolicyRow(hx.alloc, conn, runner_id) catch {
        common.internalDbError(hx.res, hx.req_id);
        return;
    } orelse {
        // Token authenticated but the row is gone (revoked + reaped) — fail
        // closed rather than 200 a phantom runner.
        hx.fail(ec.ERR_RUN_INVALID_RUNNER_TOKEN, "runner not found");
        return;
    };

    // Reconcile assigned against achievable — this beat's report when one rode
    // in, else the stored one — and carry the fresh verdict back to the host.
    const effective_cap = incoming orelse policy_row.decodeCapability(hx.alloc, row.capability_json);
    const verdict = reconcile.reconcile(row.assigned, effective_cap);
    persistVerdict(hx, conn, runner_id, incoming, row, verdict);

    // A verdict rode in on this beat. Stored after the capability write so a
    // malformed one cannot cost the beat its reconciliation, and best-effort for
    // the same reason liveness is: a self-test that fails to persist is re-run
    // by the operator, while a failed heartbeat parks real work.
    const reported = persistSelftest(hx, conn, runner_id, req);

    bumpLastSeen(hx, conn, runner_id);
    metrics_runner.touchRunnerSeen(runner_id); // in-memory liveness for /metrics
    hx.ok(.ok, protocol.HeartbeatResponse{
        .status = .ok,
        .assigned_policy = row.assigned,
        .degraded = verdict.degraded,
        .degraded_reason = verdict.reason,
        // Suppressed on the beat that just reported one: the write above cleared
        // the request, so echoing it back would ask the host to immediately
        // re-run the probe it has this second finished.
        .selftest_requested = row.selftest_requested and !reported,
    });
}

/// Store a reported verdict, if this beat carried a well-formed one. Returns
/// whether anything was written, which is what suppresses the request echo.
///
/// Silent on a malformed verdict, exactly like the capability report: a runner
/// token must not be able to fail a liveness beat by sending nonsense.
fn persistSelftest(hx: Hx, conn: *pg.Conn, runner_id: []const u8, req: *httpz.Request) bool {
    const raw = req.body() orelse return false;
    if (raw.len == 0) return false;
    const parsed = std.json.parseFromSliceLeaky(protocol.HeartbeatRequest, hx.alloc, raw, .{ .ignore_unknown_fields = true }) catch return false;
    const report = parsed.selftest orelse return false;
    switch (protocol.selftestReportRejection(report)) {
        .none => {},
        .unbounded => {
            log.warn(LOG_EVENT_SELFTEST_REJECTED, .{ .error_code = ec.ERR_RUN_SELFTEST_REPORT_INVALID, .runner_id = runner_id, .err = "verdict exceeds its bounds" });
            return false;
        },
        .all_ok_disagrees => {
            log.warn(LOG_EVENT_SELFTEST_REJECTED, .{ .error_code = ec.ERR_RUN_SELFTEST_REPORT_INVALID, .runner_id = runner_id, .err = "all_ok disagrees with the reported checks" });
            return false;
        },
    }

    const checks_json = std.json.Stringify.valueAlloc(hx.alloc, report.checks, .{}) catch return false;
    const now_ms = clock.nowMillis();
    _ = conn.exec(sql.UPDATE_RUNNER_SELFTEST, .{
        runner_id,
        checks_json,
        report.all_ok,
        report.sandbox_tier,
        report.network_policy,
        now_ms,
    }) catch |err| {
        log.warn(LOG_EVENT_SELFTEST_PERSIST_FAILED, .{ .error_code = ec.ERR_INTERNAL_DB_QUERY, .runner_id = runner_id, .err = @errorName(err) });
        return false;
    };
    log.debug("runner_selftest_reported", .{ .runner_id = runner_id, .all_ok = report.all_ok, .checks = report.checks.len });
    return true;
}

/// What one heartbeat body carried. A malformed report must never fail a
/// liveness beat: unreadable reads as "no report this beat" and the stored
/// report keeps reconciling. Only the oversize path has already responded.
const ReportParse = union(enum) { none, report: protocol.CapabilityReport, responded };

fn parseCapabilityReport(hx: Hx, req: *httpz.Request) ReportParse {
    const raw = req.body() orelse return .none;
    if (raw.len == 0) return .none;
    if (!common.checkBodySize(req, hx.res, raw, hx.req_id)) return .responded;
    const parsed = std.json.parseFromSliceLeaky(protocol.HeartbeatRequest, hx.alloc, raw, .{ .ignore_unknown_fields = true }) catch return .none;
    const report = parsed.capability_report orelse return .none;
    // An out-of-bounds controllers list is a malformed report, not a
    // persistence-amplification channel — same lenient "no report this beat".
    if (!protocol.capabilityReportBounded(report)) return .none;
    return .{ .report = report };
}

/// The row as the reconciliation needs it. Slices are re-parsed or duped into
/// the request arena — nothing borrows the query result past this function.
const PolicyRow = struct {
    assigned: ?protocol.AssignedPolicy,
    capability_json: ?[]const u8,
    stored_degraded: bool,
    stored_reason: ?[]const u8,
    /// An operator's outstanding ask, as a decided boolean rather than the raw
    /// timestamp: the reply carries "is one pending", and nothing downstream has
    /// a use for when it was made.
    selftest_requested: bool,
};

fn readPolicyRow(alloc: std.mem.Allocator, conn: *pg.Conn, runner_id: []const u8) !?PolicyRow {
    var q = PgQuery.from(conn.query(sql.SELECT_RUNNER_ASSIGNED_POLICY, .{runner_id}) catch return error.DbError);
    defer q.deinit();
    const row = (q.next() catch return error.DbError) orelse return null;
    const assigned = policy_row.decodePolicy(
        alloc,
        row.get([]const u8, 0) catch return error.DbError,
        row.get(?[]const u8, 1) catch return error.DbError,
        row.get(?[]const u8, 2) catch return error.DbError,
        row.get(i32, 3) catch return error.DbError,
        row.get(?[]const u8, 8) catch return error.DbError,
    );
    const reason_raw = row.get(?[]const u8, 5) catch return error.DbError;
    const cap_raw = row.get(?[]const u8, 6) catch return error.DbError;
    return PolicyRow{
        .assigned = assigned,
        .capability_json = if (cap_raw) |c| try alloc.dupe(u8, c) else null,
        .stored_degraded = row.get(bool, 4) catch return error.DbError,
        .stored_reason = if (reason_raw) |r| try alloc.dupe(u8, r) else null,
        .selftest_requested = (row.get(?i64, 7) catch return error.DbError) != null,
    };
}

/// Write what this beat changed: a fresh report always lands with its verdict;
/// otherwise only a moved verdict writes (the statement's guard keeps a steady
/// state write-free). Best-effort — a failed write self-heals next beat.
fn persistVerdict(hx: Hx, conn: *pg.Conn, runner_id: []const u8, incoming: ?protocol.CapabilityReport, row: PolicyRow, verdict: reconcile.Verdict) void {
    const now_ms = clock.nowMillis();
    if (incoming) |report| {
        const report_json = std.json.Stringify.valueAlloc(hx.alloc, report, .{}) catch {
            log.warn(LOG_EVENT_CAPABILITY_PERSIST_FAILED, .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .runner_id = runner_id, .err = "report allocation failed" });
            return;
        };
        _ = conn.exec(sql.UPDATE_RUNNER_CAPABILITY_AND_VERDICT, .{ runner_id, report_json, now_ms, verdict.degraded, verdict.reason }) catch |err| {
            log.warn(LOG_EVENT_CAPABILITY_PERSIST_FAILED, .{ .error_code = ec.ERR_INTERNAL_DB_QUERY, .runner_id = runner_id, .err = @errorName(err) });
        };
        log.debug("runner_capability_reported", .{ .runner_id = runner_id, .landlock = report.landlock, .seccomp = report.seccomp, .cgroup_controllers = report.cgroup_controllers.len, .bubblewrap = report.bubblewrap, .egress_enforcement = report.egress_enforcement });
    } else if (verdict.degraded != row.stored_degraded or !reasonEql(verdict.reason, row.stored_reason)) {
        _ = conn.exec(sql.UPDATE_RUNNER_VERDICT, .{ runner_id, verdict.degraded, verdict.reason, now_ms }) catch |err| {
            log.warn("verdict_persist_failed", .{ .error_code = ec.ERR_INTERNAL_DB_QUERY, .runner_id = runner_id, .err = @errorName(err) });
        };
    }
    if (verdict.degraded and !row.stored_degraded) {
        log.warn("runner_degraded", .{ .error_code = ec.ERR_EXEC_ASSIGNMENT_UNACHIEVABLE, .runner_id = runner_id, .reason = verdict.reason orelse "unspecified" });
    } else if (!verdict.degraded and row.stored_degraded) {
        log.debug("runner_degraded_cleared", .{ .runner_id = runner_id });
    }
}

fn reasonEql(a: ?[]const u8, b: ?[]const u8) bool {
    if (a == null and b == null) return true;
    if (a == null or b == null) return false;
    return std.mem.eql(u8, a.?, b.?);
}

/// Best-effort liveness bump — a DB blip must not fail the heartbeat reply.
fn bumpLastSeen(hx: Hx, conn: *pg.Conn, runner_id: []const u8) void {
    const now_ms = clock.nowMillis();
    const event_row_id = id_format.generateRunnerEventId(hx.alloc) catch |err| {
        log.warn("heartbeat_online_event_id_failed", .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .runner_id = runner_id, .err = @errorName(err) });
        bumpOnly(conn, runner_id, now_ms);
        return;
    };
    defer hx.alloc.free(event_row_id);
    _ = conn.exec(sql.HEARTBEAT_WITH_TRANSITION_EVENT, .{
        runner_id,
        now_ms,
        event_row_id,
        @tagName(protocol.RunnerEventType.runner_online),
        runner_events.META_LAST_SEEN_AT,
        protocol.RUNNER_LAST_SEEN_NEVER,
        constants.RUNNER_OFFLINE_AFTER_MS,
    }) catch |err| {
        log.warn(LOG_EVENT_HEARTBEAT_BUMP_FAILED, .{ .error_code = ec.ERR_INTERNAL_DB_QUERY, .runner_id = runner_id, .err = @errorName(err) });
        bumpOnly(conn, runner_id, now_ms);
    };
}

fn bumpOnly(conn: anytype, runner_id: []const u8, now_ms: i64) void {
    _ = conn.exec(sql.TOUCH_RUNNER_LAST_SEEN, .{ runner_id, now_ms }) catch |err| {
        log.warn(LOG_EVENT_HEARTBEAT_BUMP_FAILED, .{ .error_code = ec.ERR_INTERNAL_OPERATION_FAILED, .runner_id = runner_id, .err = @errorName(err) });
    };
}

File: rustd/crates/afd_runner/src/heartbeat.rs
//! The beat: liveness up, assignment down, and the verdict reconciled between.
//!
//! One request, four writes, and only ONE of them may fail the reply. The
//! policy read is load-bearing — the reply IS the delivery channel for an
//! operator's assignment, so a host that gets a 200 with no policy would apply
//! nothing. Every write after it is best-effort: a verdict that does not land
//! is re-derived next beat, a self-test that does not land is re-run on ask,
//! and a liveness bump that does not land costs one interval of freshness. A
//! beat that failed on any of those would park real work over a blip.
//!
//! # Why the split is a function and not a comment
//!
//! `heartbeat.zig` makes the same choice and spells it as five `catch` blocks
//! that each log and carry on. Nothing relates them, so "this write is
//! best-effort" is a property a reader reconstructs from the shape of five
//! bodies. Here it is [`best_effort`] — one function, named, with the reason in
//! its documentation, and a call site that reads as a decision rather than as
//! an omission.

use afd_core::clock::UnixMillis;
use afd_core::error_code;
use afd_core::id::{ENTROPY_LEN, Uuid7};
use afd_core::timing::RUNNER_OFFLINE_AFTER_MS;
use afd_observability::producers;
use afd_wire::runner::{CapabilityReport, HeartbeatRequest, SelftestReport};
use sqlx::{Executor as _, PgConnection, Row as _};

use crate::bounds;
use crate::error::{Error, Result, query};
use crate::policy::{AssignmentColumns, StoredVerdict, capability};
use crate::reconcile::{Verdict, reconcile};
use crate::spelling::render_list;
use crate::sql;
use crate::sql::runner::Bound;
use crate::store::Runners;

/// The statement name a policy-read failure carries.
const CONTEXT_POLICY_READ: &str = "runner policy read";

/// The scoped events a best-effort write reports itself under.
const EVENT_VERDICT_WRITE: &str = "verdict_persist_failed";
const EVENT_SELFTEST_WRITE: &str = "selftest_persist_failed";
const EVENT_LIVENESS_WRITE: &str = "heartbeat_bump_failed";

/// A beat that carried nothing — the shape a body-less or unreadable heartbeat
/// resolves to.
///
/// A runner token must not be able to fail a liveness beat by sending nonsense,
/// so the caller substitutes this rather than refusing the request; the stored
/// report keeps reconciling and the host is told its liveness landed.
pub const NO_REPORT: HeartbeatRequest<'static> = HeartbeatRequest {
    capability_report: None,
    selftest: None,
};

/// What one beat resolved to, for the reply the host reads.
#[derive(Debug)]
pub struct Beat {
    /// The assignment as stored, for the caller to decode and echo.
    pub assignment: AssignmentColumns,
    /// The verdict this beat reconciled.
    pub verdict: Verdict,
    /// Whether an operator's self-test request is still outstanding.
    ///
    /// Suppressed on the beat that just reported one: the write cleared the
    /// ask, so echoing it back would tell the host to immediately re-run the
    /// probe it has this second finished.
    pub selftest_requested: bool,
}

/// The policy row, as the reconciliation needs it.
#[derive(Debug)]
struct PolicyRow {
    assignment: AssignmentColumns,
    stored: StoredVerdict,
    capability_report_json: Option<String>,
    selftest_requested: bool,
}

impl Runners {
    /// Records a beat and answers what the host must apply.
    ///
    /// # Errors
    /// Reports a datastore that would not answer, a statement Postgres refused,
    /// and — as its own kind — a token that authenticated against a row which
    /// has since been reaped. Nothing else fails a beat: see the module
    /// documentation for which writes are best-effort and why.
    pub async fn heartbeat(
        &self,
        runner: &Uuid7,
        beat: &HeartbeatRequest<'_>,
        now: UnixMillis,
    ) -> Result<Beat> {
        let mut connection = self.pool().acquire().await?;
        let row = self.policy_row(&mut connection, runner).await?;

        // A report past its bounds is not a report. Same lenient answer as an
        // unreadable body: the stored one keeps reconciling.
        let incoming = beat
            .capability_report
            .as_ref()
            .filter(|report| bounds::capability_within_bounds(report));
        let stored = capability(row.capability_report_json.as_deref());
        let assigned = row.assignment.decode();
        let verdict = reconcile(assigned.as_ref(), incoming.or(stored.as_ref()));

        persist_verdict(&mut connection, runner, incoming, &row.stored, verdict, now).await;
        // After the capability write, so a malformed verdict cannot cost the
        // beat its reconciliation.
        let reported = persist_selftest(&mut connection, runner, beat.selftest.as_ref(), now).await;
        self.bump_liveness(&mut connection, runner, now).await;
        // The gauge's only input. Liveness is a Postgres row a collection
        // callback cannot read — it is a network round trip, and the SDK
        // collects on a thread that must not make one — so the beat that
        // writes the row publishes the same instant for the gauge to serve.
        producers::fleet::runner::seen(runner.as_str(), now.as_millis());

        Ok(Beat {
            selftest_requested: row.selftest_requested && !reported,
            assignment: row.assignment,
            verdict,
        })
    }

    /// Reads the assignment, the stored report, and the prior verdict.
    ///
    /// The one read that fails loud: the reply is how an assignment reaches the
    /// host, so answering 200 with no policy would tell a runner to apply
    /// nothing.
    async fn policy_row(&self, connection: &mut PgConnection, runner: &Uuid7) -> Result<PolicyRow> {
        let found = sqlx::query(sql::runner::SELECT_RUNNER_ASSIGNED_POLICY)
            .bind(runner.as_str())
            .fetch_optional(&mut *connection)
            .await
            .map_err(query(CONTEXT_POLICY_READ))?;
        // Fail closed rather than beat a phantom runner: the token is real and
        // the enrolment is gone, so the host must be re-enrolled.
        let row = found.ok_or_else(|| Error::RunnerVanished)?;

        let column = query(CONTEXT_POLICY_READ);
        Ok(PolicyRow {
            assignment: AssignmentColumns {
                sandbox_tier: row.try_get(0).map_err(&column)?,
                network_policy: row.try_get(1).map_err(&column)?,
                registry_allowlist_json: row.try_get(2).map_err(&column)?,
                worker_count: row.try_get(3).map_err(&column)?,
                extra_binds_json: row.try_get(8).map_err(&column)?,
            },
            stored: StoredVerdict {
                degraded: row.try_get(4).map_err(&column)?,
                reason: row.try_get(5).map_err(&column)?,
            },
            capability_report_json: row.try_get(6).map_err(&column)?,
            // A decided boolean rather than the raw instant: the reply carries
            // "is one pending", and nothing downstream has a use for when it
            // was made.
            selftest_requested: row.try_get::<Option<i64>, _>(7).map_err(&column)?.is_some(),
        })
    }

    /// Bumps liveness, emitting a transition event only on a real transition.
    ///
    /// Falls back to the liveness-only statement when the event identifier
    /// cannot be minted or the combined statement is refused: a beat that
    /// writes liveness without its audit row is a quieter loss than one that
    /// writes nothing, because the fleet read stays true.
    async fn bump_liveness(&self, connection: &mut PgConnection, runner: &Uuid7, now: UnixMillis) {
        let landed = match self.event_id(now) {
            Ok(event_id) => {
                let millis = now.as_millis();
                let with_event = sqlx::query(sql::runner::HEARTBEAT_WITH_TRANSITION_EVENT)
                    .bind(runner.as_str())
                    .bind(millis)
                    .bind(event_id.as_str())
                    .bind(sql::event_type::RUNNER_ONLINE)
                    .bind(sql::meta::LAST_SEEN_AT)
                    .bind(sql::LAST_SEEN_NEVER)
                    .bind(RUNNER_OFFLINE_AFTER_MS);
                best_effort(with_event, connection, EVENT_LIVENESS_WRITE, runner).await
            }
            Err(error) => {
                report(EVENT_LIVENESS_WRITE, runner, &error);
                false
            }
        };
        if !landed {
            let touch = sqlx::query(sql::runner::TOUCH_RUNNER_LAST_SEEN)
                .bind(runner.as_str())
                .bind(now.as_millis());
            best_effort(touch, connection, EVENT_LIVENESS_WRITE, runner).await;
        }
    }

    /// Draws the identifier a transition event is written under.
    fn event_id(&self, now: UnixMillis) -> Result<Uuid7> {
        let mut bytes = [0u8; ENTROPY_LEN];
        self.entropy().fill(&mut bytes)?;
        Ok(Uuid7::encode(now, bytes)?)
    }
}

/// Writes what this beat changed about the verdict.
///
/// A fresh report always lands with its verdict; otherwise only a MOVED verdict
/// writes, and the statement's own guard makes a steady state write nothing at
/// all. The `differs_from` check ahead of it saves the round trip that guard
/// would otherwise cost on every beat of every idle host.
async fn persist_verdict(
    connection: &mut PgConnection,
    runner: &Uuid7,
    incoming: Option<&CapabilityReport<'_>>,
    stored: &StoredVerdict,
    verdict: Verdict,
    now: UnixMillis,
) {
    let millis = now.as_millis();
    if let Some(report) = incoming {
        let report_json = serde_json::to_string(report).unwrap_or_else(|_unreachable| {
            // Unreachable for this shape — booleans and a string list — and an
            // empty object is the honest degradation: it stores "reported
            // nothing" rather than a half-written report.
            "{}".to_owned()
        });
        let write = sqlx::query(sql::runner::UPDATE_RUNNER_CAPABILITY_AND_VERDICT)
            .bind(runner.as_str())
            .bind(report_json)
            .bind(millis)
            .bind(verdict.is_degraded())
            .bind(verdict.reason());
        best_effort(write, connection, EVENT_VERDICT_WRITE, runner).await;
    } else if stored.differs_from(verdict) {
        let write = sqlx::query(sql::runner::UPDATE_RUNNER_VERDICT)
            .bind(runner.as_str())
            .bind(verdict.is_degraded())
            .bind(verdict.reason())
            .bind(millis);
        best_effort(write, connection, EVENT_VERDICT_WRITE, runner).await;
    }
    announce(runner, stored, verdict);
}

/// Stores a reported verdict, and answers whether anything was written.
///
/// Silent on a malformed verdict, exactly like the capability report: a runner
/// token must not be able to fail a liveness beat by sending nonsense.
async fn persist_selftest(
    connection: &mut PgConnection,
    runner: &Uuid7,
    reported: Option<&SelftestReport<'_>>,
    now: UnixMillis,
) -> bool {
    let Some(report) = reported else {
        return false;
    };
    if let Err(rejection) = bounds::accept(report) {
        let code = error_code::INVALID_REQUEST.as_str();
        let id = runner.as_str();
        let reason = rejection.to_string();
        tracing::warn!(
            error_code = code,
            runner_id = id,
            reason,
            event = "selftest_verdict_refused",
            "self-test verdict refused; the beat still counts as liveness"
        );
        return false;
    }

    let checks_json = render_list(&report.checks);
    let write = sqlx::query(sql::runner::UPDATE_RUNNER_SELFTEST)
        .bind(runner.as_str())
        .bind(checks_json)
        .bind(report.all_ok)
        .bind(report.sandbox_tier.as_ref())
        .bind(report.network_policy.as_ref())
        .bind(now.as_millis());
    best_effort(write, connection, EVENT_SELFTEST_WRITE, runner).await
}

/// Runs a write whose failure must not fail the beat, and says whether it
/// landed.
///
/// The single place the best-effort rule is expressed. Every caller of it is a
/// write the next beat re-derives, so a failure is logged at `warn` — visible
/// to an operator watching one host, invisible to the reply.
async fn best_effort(
    statement: Bound<'_>,
    connection: &mut PgConnection,
    event: &'static str,
    runner: &Uuid7,
) -> bool {
    match connection.execute(statement).await {
        Ok(_landed) => true,
        Err(error) => {
            let code = error_code::INTERNAL_DB_QUERY.as_str();
            let id = runner.as_str();
            let reason = error.to_string();
            tracing::warn!(error_code = code, runner_id = id, reason, event);
            false
        }
    }
}

/// Reports a best-effort step that could not even be attempted.
fn report(event: &'static str, runner: &Uuid7, error: &Error) {
    let code = error_code::INTERNAL_OPERATION_FAILED.as_str();
    let id = runner.as_str();
    let reason = error.to_string();
    tracing::warn!(error_code = code, runner_id = id, reason, event);
}

/// Says a verdict CHANGED, and only when it changed.
///
/// A degradation is an operator event — a host that will not take work — so it
/// is `warn` on the transition and silent on every beat after it. Logging the
/// state rather than the transition would put one line per host per ten seconds
/// into the log and hide the moment it happened.
fn announce(runner: &Uuid7, stored: &StoredVerdict, verdict: Verdict) {
    let id = runner.as_str();
    match (stored.degraded, verdict.reason()) {
        (false, Some(reason)) => {
            let code = error_code::INTERNAL_OPERATION_FAILED.as_str();
            tracing::warn!(
                error_code = code,
                runner_id = id,
                reason,
                event = "runner_degraded",
                "runner degraded — it will not be assigned work until this is fixed"
            );
        }
        (true, None) => tracing::debug!(runner_id = id, event = "runner_recovered"),
        _steady => {}
    }
}
